add pretty print key, #2928

veteranlu opened the pull request:
Key organized as follow:
- /Local
  - /Store
  - /RangeID
  - /Range
- /System
  - /Meta1
  - /Meta2
  - /StatusStore
  - /StatusNode
- /Table

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3055)

<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=3055) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

petermattis commented on keys/keys_test.go:
```diff
@@ -20,6 +20,7 @@ package keys
 import (
```
Looks like you made this file executable (mode bits changed from 644 to 755). I assume that was accidental. If not, please justify the change.

veteranlu replied:
I don't know why its mode bits modified to 755, I will change the mode bits back to 644.

petermattis commented on keys/keys.go:
```diff
@@ -369,3 +371,251 @@ func Range(ba roachpb.BatchRequest) roachpb.RSpan {
+			//{name: "Status", prefix: StatusPrefix},
+		}},
+		{name: "User", start: TableDataPrefix, end: nil, entries: []dictEntry{
+			{name: "TableData", prefix: TableDataPrefix, ppFunc: decodeKeyPrint},
```
All user data is table data and that is unlikely to change any time soon. Can we shorten from `/User/TableData` to `/Table`?

veteranlu replied:
ok

petermattis commented on keys/keys.go:
```diff
@@ -21,8 +21,10 @@ package keys
+	"time"
 
 	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/sql/parser"
```
This is an awkward import to add. I think it can be avoided by avoiding usage of the `parser.Datum` types in `decodeKeyPrint`. They aren't doing much.

petermattis commented on keys/keys.go:
```diff
@@ -369,3 +371,251 @@ func Range(ba roachpb.BatchRequest) roachpb.RSpan {
+		switch encoding.PeekType(key) {
+		case encoding.Null:
+			key, _ = encoding.DecodeIfNull(key)
+			d = parser.DNull
```
Rather than initializing `d`, you could avoid the `parser` dependency by outputting the value:

```
  fmt.Fprintf(&buf, "/NULL")
```

veteranlu replied:
sure

petermattis commented on keys/keys.go:
```diff
@@ -369,3 +371,251 @@ func Range(ba roachpb.BatchRequest) roachpb.RSpan {
+		case encoding.Bytes:
+			var s string
+			key, s, err = encoding.DecodeString(key, nil)
+			d = parser.DString(s)
```
This is the only real tricky case to handle due to escaping, but I think something like the following is acceptable:

```
  fmt.Fprintf(&buf, "/%q", s)
```

veteranlu replied:
ok

petermattis commented:
Thanks for tackling this. Pretty close to what I was imagining when I filed the issue. 

I think we'll want `roachpb.Key.String()` to perform pretty-printing, otherwise we have hundreds of log messages to change. But this is complicated by `keys` depending on `roachpb` so `roachpb` cannot depend on `keys`. I don't have an answer for how to fix this, but we need to figure something out.

tamird commented:
I think I'd like to see `sql.prettyKey` replaced with this implementation, just to make sure it all fits together nicely.

veteranlu commented:
@petermattis I was supposed to add `PrettyString` to `roachpb.Key.`, but I met the problem of dependence, `keys.PrettyPrint` is temporary, I will try to figure out the dependence problem.

veteranlu commented:
@tamird I would like to add test case to cover more scenarios.

petermattis commented:
Nice.

One hack to get this hooked into `roachpb.Key.String()` would be to add a `var PrettyPrintKey func(Key)` global to the `roachpb` package which the keys package can fill in via an init function:

```
package keys
...
func init() {
  roachpb.PrettyPrintKey = PrettyPrint
}
```

Then `roachpb.Key.String()` would become:

```
func (k Key) String() string {
  if PrettyPrintKey != nil {
    return PrettyPrintKey(k)
  }
  return fmt.Sprintf("%q", []byte(k))
}
```

You'd have to do something similar for `roachpb.Key.Format()`. Sort of ugly, but it gets rid of the circular dependency. 
<!-- Reviewable comment -K2lta2jzGGexlDCdqct:-889380266 -->

---

Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

<sup>**[keys/keys.go, line 415 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2lsCeic020XyRydQ35)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L415)):</sup>
`s/key.String()/key/g` (fmt will take care of calling `String()` for you).
<!-- Reviewable comment -K2lsCeic020XyRydQ36:318454967 -->

---

<sup>**[keys/keys.go, line 416 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2lsTtpzHzaLovwjzEP)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L416)):</sup>
Similar to below: `return fmt.Sprintf("/err<%v:%s>", err, key)`
<!-- Reviewable comment -K2lsTtpzHzaLovwjzEQ:-1951263345 -->

---

<sup>**[keys/keys.go, line 420 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2lsMFzRs0If-HGaz73)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L420)):</sup>
Instead of using `buf`, couldn't you just: `return fmt.Sprintf("/logIndex:%d", logIndex)`
<!-- Reviewable comment -K2lsMFzRs0If-HGaz74:218731492 -->

---

<sup>**[keys/keys.go, line 433 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2lsccFg64SH2Z5_PDL)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L433)):</sup>
Is it necessary to predeclare `i` and `err` here. I would think the `:=` syntax would work:

```
  key, i, err := encoding.DecodeVarint(key)
```
<!-- Reviewable comment -K2lsccFg64SH2Z5_PDM:-928918302 -->

---

<sup>**[keys/keys_test.go, line 385 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2lt2cEedzbGZ0W1y9t)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys_test.go#L385)):</sup>
It can be nice to use backquoted strings so that you don't have quote your double-quotes. For example, the expected string above could be replaced with:

```
`/Local/Range/"ok"`
```
<!-- Reviewable comment -K2lt2cEedzbGZ0W1y9u:-1827550518 -->

---

<sup>**[keys/keys_test.go, line 418 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2ltKwgbXYyYwpCVDri)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys_test.go#L418)):</sup>
`if test.exp != keyInfo`?
<!-- Reviewable comment -K2ltKwgbXYyYwpCVDrj:1897106591 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055)
<!-- Sent from Reviewable.io -->

petermattis commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055)
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

<sup>**[keys/keys.go, line 389 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mEjf8SnumaHpjr2II)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L389)):</sup>
FYI this is going to change real soon. cc @tschottdorf
<!-- Reviewable comment -K2mEjf8SnumaHpjr2IJ:-64303185 -->

---

<sup>**[keys/keys.go, line 426 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mFS9bdR8ii_AePESH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L426)):</sup>
there's a bunch of overuse of `bytes.Buffer` here
<!-- Reviewable comment -K2mFS9bdR8ii_AePESI:-2036125044 -->

---

<sup>**[keys/keys.go, line 435 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mFBjWLr0NQU6harQY)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L435)):</sup>
s/String()//, fmt calls `String()` for you
<!-- Reviewable comment -K2mFBjXNFBEjzaOSdXu:468442875 -->

---

<sup>**[keys/keys.go, line 442 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mFarARqUBrBtSIeQ0)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L442)):</sup>
allocated on every call?
<!-- Reviewable comment -K2mFarARqUBrBtSIeQ1:37992985 -->

---

<sup>**[keys/keys.go, line 461 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mF5jsHrHgwFgIuZol)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L461)):</sup>
how come everything references suffixes while this checks the prefix?
<!-- Reviewable comment -K2mF5jsHrHgwFgIuZom:-2007725683 -->

---

<sup>**[keys/keys.go, line 486 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mFZJmvHmwb4lfjOoK)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L486)):</sup>
looks like this gets allocated on every call?
<!-- Reviewable comment -K2mFZJmvHmwb4lfjOoL:-149110067 -->

---

<sup>**[keys/keys.go, line 573 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K2mEX04VVkx9g1VHZbh)** ([raw file](https://github.com/cockroachdb/cockroach/blob/4fa467483b14174eddcf20043cc18ebeb55f6a22/keys/keys.go#L573)):</sup>
This slice looks like it gets allocated on every call to PrettyPrint - it should probably be allocated once outside the function.
<!-- Reviewable comment -K2mEX04VVkx9g1VHZbi:-1385701238 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055)
<!-- Sent from Reviewable.io -->

veteranlu commented:
@petermattis maybe it's the best one of the three way.
http://stackoverflow.com/questions/29271440/registering-packages-in-go-without-cyclic-dependency

`roachpb.Key.Format()` shoule not be changed, it will busy loop if we use `PrettyPrint` for `roachpb.Key.Format()`.

petermattis commented:
FYI, `roachpb.Key.Format()` was added because we aren't strict about using the `%s` format for keys everywhere. Can you explain why you think having `roachpb.Key.Format()` call out to `PrettyPrint()` will cause a busy loop? I imagine a bit of casting to `[]byte` in the `PrettyPrint()` implementation can get rid of that problem.

veteranlu commented:
@petermattis casting to `[]byte` in `PrettyPrint()` can get rid of that problem.

petermattis commented:
LGTM

Do you mind squashing the commits?
<!-- Reviewable comment -K3FiZ9QnQfERKcSFcb_:-92117184 -->

---

Reviewed 1 of 2 files at r3.
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[cli/cli_test.go, line 228 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K3FhcBGtWBfCx3YvLaH)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae22b0d0dfd7ed8e4b41fbce466a1308ca0583c/cli/cli_test.go#L228)):</sup>
That's not quite right. `"\xff\xff"` is the infinity marker which is larger than any other key. It isn't part of the `"/Table"` space. Can we display it as `"/Infinity"` or `"/Max"`?
<!-- Reviewable comment -K3FhcBGtWBfCx3YvLaI:-552049077 -->

---

<sup>**[keys/keys.go, line 567 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055#-K3FiuaQVtL3KXzHgebv)** ([raw file](https://github.com/cockroachdb/cockroach/blob/3ae22b0d0dfd7ed8e4b41fbce466a1308ca0583c/keys/keys.go#L567)):</sup>
This comment appears incomplete. Also, the first part needs to be adjusted slightly: `PrettyPrint prints the key in a human readable format, organized as:`
<!-- Reviewable comment -K3FiuaRawr25nnnM1FA:660644304 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3055)
<!-- Sent from Reviewable.io -->

veteranlu commented:
@petermattis Can u help me review the source code?

petermattis commented:
@veteranlu I just sent some additional comments. Small stuff. This is almost ready to merge.

veteranlu committed:
add key pretty print, #2928

veteranlu commented:
@petermattis it's pleasure to get your advice. I modified the code as your advice, and more, I refactor the source code.
* moved the code about `PrettyPrint` to file printer.go and printer_test.go, it should make sence.
* and add more specific description for `SequenceCacheKey`.
* add /Max, and /Min as a couple.

petermattis commented:
LGTM. Thanks for tackling this.
